src/core/ext/xds/xds_client.cc

```diff
@@ -522,27 +518,53 @@ void XdsClient::ChannelState::UnsubscribeLocked(const XdsResourceType* type,
   }
 }
 
-void XdsClient::ChannelState::OnConnectivityStateChange(absl::Status status) {
+void XdsClient::ChannelState::OnConnectivityFailure(absl::Status status) {
   {
     MutexLock lock(&xds_client_->mu_);
-    OnConnectivityStateChangeLocked(std::move(status));
+    SetChannelStatusLocked(std::move(status));
   }
   xds_client_->work_serializer_.DrainQueue();
 }
 
-void XdsClient::ChannelState::OnConnectivityStateChangeLocked(
-    absl::Status status) {
-  if (!shutting_down_) {
-    // Notify all watchers of error.
-    gpr_log(GPR_INFO,
-            "[xds_client %p] xds channel for server %s in "
-            "state TRANSIENT_FAILURE: %s",
-            xds_client(), server_.server_uri.c_str(),
-            status.ToString().c_str());
-    xds_client_->NotifyOnErrorLocked(absl::UnavailableError(
-        absl::StrCat("xds channel in TRANSIENT_FAILURE, connectivity error: ",
-                     status.ToString())));
+void XdsClient::ChannelState::SetChannelStatusLocked(absl::Status status) {
+  if (shutting_down_) return;
+  status = absl::Status(
+      status.code(), absl::StrCat("xDS channel for server ", server_.server_uri,
+                                  ": ", status.message()));
+  gpr_log(GPR_INFO, "[xds_client %p] %s", xds_client(),
```

Comment:
shouldn't this be protected by a trace flag?

Reply:
Nope.  This is a message we want to log even if tracing is disabled, because it's an unusual situation that is important to log.

Note that even before this PR, this message was already being logged without a tracer enabled when the channel reported TRANSIENT_FAILURE.  The only thing new in this PR is that it will now also be logged when the ADS stream is closed.